### PR TITLE
Provide exactly once semantics

### PR DESCRIPTION
**Pull request by derekcollison**
This provides two mechanisms for exactly once.

1. Deduplication for messages that are published multiple times. We use headers to denote a Msg-Id per message. These are kept for a certain window of time.
2. When you ack a message that has been consumed you can receive an ack to your ack. If that ack is received you will not receive the message again.

Signed-off-by: Derek Collison <derek@nats.io>

/cc @nats-io/core

### CODE REVIEW

File: server/stream.go [hunk 1]
```diff
@@ -597,6 +623,91 @@ func (mset *Stream) setupStore(fsCfg *FileStoreConfig) error {
...
+	mset.ddmap[dde.id] = dde
+	mset.ddarr = append(mset.ddarr, dde)
+	mset.mu.Unlock()
+}
+
+// Will return the value for the header denoted by key or nil if it does not exists.
+// This function ignores errors and tries to achieve speed and no additional allocations.
+func getHdrVal(key string, hdr []byte) []byte {
+	var value []byte
+	for i := bytes.Index(hdr, []byte(key)) + len(key) + 2; i > 0 && i < len(hdr); i++ {
```
Comment (kozlovic) [inline server/stream.go:697]: I am confused. If the headers does not contain the key you are looking for, bytes.Index() would return -1. Why even search then?

File: server/stream.go [hunk 2]
```diff
@@ -613,6 +724,21 @@ func (mset *Stream) processInboundJetStreamMsg(_ *subscription, pc *client, subj
 	name := mset.config.Name
 	maxMsgSize := int(mset.config.MaxMsgSize)
 	numConsumers := len(mset.consumers)
+
+	// Process msgId if we have headers.
+	var msgId string
+	if pc != nil && pc.pa.hdr > 0 {
+		msgId = getMsgId(msg[:pc.pa.hdr])
```
Comment (kozlovic) [inline server/stream.go:731]: Having headers does not guarantee that the JSPubId will be present, so should check msg Id for being say empty to indicate that there is none.

File: server/stream.go (hunk 1 above)
Comment (derekcollison) [reply server/stream.go:697]: I had the loop setup early on such that it would bail early, will tweak it to be better and bail early again.

File: server/stream.go (hunk 2 above)
Comment (derekcollison) [reply server/stream.go:731]: msgId will be "" and the rest of the code will do the right thing.

File: server/stream.go (hunk 1 above)
Comment (derekcollison) [reply server/stream.go]: Should be fixed.

File: server/stream.go [hunk 3]
```diff
@@ -597,6 +623,95 @@ func (mset *Stream) setupStore(fsCfg *FileStoreConfig) error {
...
+	index := bytes.Index(hdr, []byte(key))
+	if index < 0 {
+		return nil
+	}
+	var value []byte
+	for i := index + len(key) + 2; i > 0 && i < len(hdr); i++ {
+		if hdr[i] == '\r' && i < len(hdr)-1 && hdr[i+1] == '\n' {
+			break
+		}
+		value = append(value, hdr[i])
```
Comment (kozlovic) [inline server/stream.go:705]: Should we return a slice instead of appending bytes?

File: server/stream.go (hunk 2 above)
Comment (kozlovic) [reply server/stream.go:735]: Sure, but if we could save a map lookup, why not?

File: server/stream.go (hunk 3 above)
Comment (derekcollison) [reply server/stream.go:705]: Actually opposite, do not want or actually can reference a blind read buffer that could be large.

**Comment by derekcollison** [comment]
We do not do the map lookup if we pass in ""

File: server/stream.go (hunk 2 above)
Comment (derekcollison) [reply server/stream.go:735]: We don't do a map lookup if we have an empty key.

File: server/stream.go [hunk 4]
```diff
@@ -48,6 +49,7 @@ type StreamConfig struct {
 	Replicas     int             `json:"num_replicas"`
 	NoAck        bool            `json:"no_ack,omitempty"`
 	Template     string          `json:"template_owner,omitempty"`
+	Duplicates   time.Duration   `json:"duplicates,omitempty"`
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/stream.go:52]: How about `dup_window`?

File: server/stream.go [hunk 5]
```diff
@@ -79,11 +81,27 @@ type Stream struct {
...
 	config    StreamConfig
 	created   time.Time
+	ddmap     map[string]*ddentry
+	ddarr     []*ddentry
+	ddindex   int
+	ddtmr     *time.Timer
 }
 
+// JSPubId is used for identifying published messages and performing de-duplication.
+const JSPubId = "Msg-Id"
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/stream.go:91]: I don't think we make it a config item. And I went back and forth on it. I did not like `Js-Msg-Id` but we could go with that.

File: server/stream.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ripienaar) [reply server/stream.go:52]: ok

File: server/stream.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ripienaar) [reply server/stream.go:91]: ok, but we need to accept it case insensitive as per http spec

File: server/stream.go (hunk 3 above)
Comment (kozlovic) [reply server/stream.go:705]: Right. I was just wondering if this would be more of the caller to decide if a copy is required or not. But I guess that if a message ID is present it will always need to be stored (except in error conditions, but that's ok), so we better make the copy here.

File: server/stream.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/stream.go:91]: Headers have a canonical name that is the way it is above. Uppercase for each word, words can be separated by `-`

File: server/stream.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/stream.go:52]: Went with `duplicate_window`

File: server/stream.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (kozlovic) [reply server/stream.go:91]: Yes, so for the storage in the map, we need to decide on converting to lower or upper prior that.
By the way, this is just stored in memory as of now or did I miss it? Meaning would that survive a server restart?

[the remark this answers is not in the document]
Comment (derekcollison) [reply server/stream.go:91]: Good point, it is saved but not in that form per se, would need to re-populate. Will make that test later this am and fix. Good catch.

[the remark this answers is not in the document]
Comment (kozlovic) [reply server/stream.go:91]: About case sensitive: our own ADR (https://github.com/nats-io/nats-server/blob/master/doc/adr/0004-nats-headers.md) says:
```
Each header field consists of a case-insensitive field name followed by a colon (:), optional leading whitespace, the field value, and optional trailing whitespace.
```

[the remark this answers is not in the document]
Comment (kozlovic) [reply server/stream.go:91]: Or: https://tools.ietf.org/html/rfc7230#section-3.2
```
Each header field consists of a case-insensitive field name followed
   by a colon (":"), optional leading whitespace, the field value, and
   optional trailing whitespace.
```

[the remark this answers is not in the document]
Comment (derekcollison) [reply server/stream.go:91]: https://golang.org/pkg/net/textproto/#CanonicalMIMEHeaderKey

[the remark this answers is not in the document]
Comment (kozlovic) [reply server/stream.go:91]: Those headers will come from clients and are not necessarily Go. If you think they must be set like that, then the ADR should be updated...

[the remark this answers is not in the document]
Comment (derekcollison) [reply server/stream.go:91]: Yes they should take that form, do not want to do multiple searches, already had to re-architect because generic hoisting to http style headers was too slow.

[the remark this answers is not in the document]
Comment (ripienaar) [reply server/stream.go:91]: I think its important that if we're going to tell a http header story then its a minimum wtf/minute story, so should be case insensitive.

[the remark this answers is not in the document]
Comment (kozlovic) [reply server/stream.go:91]: Agreed. Some client libraries may use some http libraries that will adhere to the http spec.

[the remark this answers is not in the document]
Comment (derekcollison) [reply server/stream.go:91]: All client libs that do HTTP do the canonical form and that is what we will search for an require in the server. Too expensive and not worth it to lowercase copies of headers. Purposely trying to avoid additional allocations and copies etc.

File: server/stream.go [hunk 6]
```diff
@@ -173,11 +185,46 @@ func (a *Account) AddStreamWithStore(config *StreamConfig, fsConfig *FileStoreCo
 	mset.pubAck = append(mset.pubAck, OK...)
 	mset.pubAck = append(mset.pubAck, fmt.Sprintf(" {\"stream\": %q, \"seq\": ", cfg.Name)...)
 
+	// Setup subscriptions
+	if err := mset.subscribeToStream(); err != nil {
```
Comment (kozlovic) [inline server/stream.go:189]: As soon as you subscribe, you may receive messages that would possibly not be de-duped, no? In other words, should (and is it possible) the mset.rebuildDedupe() function be called before this one?

Comment (derekcollison) [reply server/stream.go:189]: Geez, that is why I was moving it and moved it not far enough.. Will fix.

Comment (derekcollison) [reply server/stream.go:192]: ok moved and force pushed.

File: server/stream.go [hunk 7]
```diff
@@ -318,6 +365,9 @@ func checkStreamCfg(config *StreamConfig) (StreamConfig, error) {
 	if cfg.MaxConsumers == 0 {
 		cfg.MaxConsumers = -1
 	}
+	if cfg.Duplicates == 0 {
```
Comment (ripienaar) [inline server/stream.go:368]: this is allowing negative times, is that intentional? For documentation whats the intended behaviour of negative times? I think we should fail to create the stream in that case if not intentional

Comment (ripienaar) [reply server/stream.go:368]: Also should we set a upper limit maybe? It would be bad if someone set this to a year

Comment (ripienaar) [reply server/stream.go:368]:
```
$ nats req -H Msg-Id:1 in.x hello
11:55:24 Sending request on [in.x]
11:55:24 Received on [_INBOX.YyMIz3ytiJtzdus9c8kbXj]
+OK {"stream": "X", "seq": 1}
```

What do you think about adding a field to the response in the case of a duplicate ie `+OK {"stream": "X", "seq": 1, "dupe": true}` or similar?

Comment (derekcollison) [reply server/stream.go:368]: These are good suggestions. Will add the test for negative and also add in `duplicate` field. What feels like a good upper bound? I agree but do not know what the right number is TBH. For now I may suggest we keep open, we can add one later.

I can limit if they have a limit on the stream itself as well.

Comment (ripienaar) [reply server/stream.go:368]: difficult to say what a good upper limit is tbh, but surely a day should be plenty, hard to be sure what the use cases for this is tbh
